### PR TITLE
Kyber: Add option to build Kyber API

### PR DESCRIPTION
**Pull request by SparkiDev**
# Description

wolfSSL Kyber implementation not included.
Added tests and benchmarking.

# Testing

Tested with wolfSSL implementation.
Tested that not copying in wolfSSL implementation fails build.
Just using --enable-kyber will fail as there is no implementation at this time.

# Checklist

 - [x] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -20689,6 +20693,13 @@ const char* wolfSSL_get_curve_name(WOLFSSL* ssl)
 #elif defined(HAVE_PQM4)
         case WOLFSSL_KYBER_LEVEL1:
             return "KYBER_LEVEL1";
+#elif defined(WOLFSSL_WC_KYBER)
+        case WOLFSSL_KYBER_LEVEL1:
+            return "KYBER_LEVEL1";
+        case WOLFSSL_KYBER_LEVEL3:
+            return "KYBER_LEVEL3";
+        case WOLFSSL_KYBER_LEVEL5:
+            return "KYBER_LEVEL5";
```
Comment (anhu) [inline src/ssl.c:20702]: Should each of the case statements be guarded by `WOLFSSL_NO_KYBERxxx` ? 
Should there be `90s` cases as well?

**Review by anhu — COMMENTED** [review]
Does HAVE_PQC need to be defined in configure.ac?

**Comment by anhu** [comment]
I copied over the kyber implementation and did : 

```
./configure --enable-kyber
```
Got the following output which I found strange: 

```
   * KYBER:                      yes
   * KYBER wolfSSL impl:         no
```

**Comment by anhu** [comment]
I copied over the kyber implementation and did : 

```
./configure --enable-kyber
```
Got the following output which I found strange: 

```
   * KYBER:                      yes
   * KYBER wolfSSL impl:         no
```

**Comment by anhu** [comment]
I understand what I did wrong. I needed `./configure --enable-kyber=wolfssl,yes`

This lets everything compile but at link time there undefined references to `wc_KyberKey_*` in benchmark app.

File: wolfssl/wolfcrypt/kyber.h [hunk 2]
```diff
@@ -0,0 +1,228 @@
+/* kyber.h
```
Comment (anhu) [inline wolfssl/wolfcrypt/kyber.h:1]: Do we need some stuff for 90S as well?

**Review by anhu — CHANGES_REQUESTED** [review]

File: wolfssl/wolfcrypt/kyber.h [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+#define KYBER_MAX_CIPHER_TEXT_SIZE  KYBER512_CIPHER_TEXT_SIZE
+#endif
+
+
+
+enum {
+    /* Types of Kyber keys. */
+    KYBER512  = 0,
+    KYBER768  = 1,
+    KYBER1024 = 2,
```
Comment (anhu) [inline wolfssl/wolfcrypt/kyber.h:161]: do we need 90S here?

**Comment by JacobBarthelmeh** [comment]
> Having trouble with missing symbols at link stage. Showed commands used in previous message.

I see this too. Note that --enable-static --disable-shared allows it to link successfully, so appears to be a problem when dynamically linking.

File: configure.ac [hunk 4]
```diff
@@ -963,6 +963,78 @@ then
...
+do
+  case $v in
+  yes | all)
+    ENABLED_KYBER512=yes
+    ENABLED_KYBER768=yes
+    ENABLED_KYBER1024=yes
+    ;;
+  no)
+    ;;
+  wolfssl)
```
Comment (JacobBarthelmeh) [inline configure.ac:987]: I think this should default to the wolfssl implementation. When using it the only config needed for a basic build would be --enable-kyber and if overriding the basic build then need the extra arguments.

File: src/tls.c [hunk 5]
```diff
@@ -7150,7 +7183,120 @@ static void findEccPqc(int *ecc, int *pqc, int group)
...
+
+    findEccPqc(&ecc_group, &oqs_group, kse->group);
+    ret = kyber_id2type(oqs_group, &type);
+    if (ret == NOT_COMPILED_IN) {
+        WOLFSSL_MSG("Invalid Kyber algorithm specified.");
+        ret = BAD_FUNC_ARG;
+    }
+
+    if (ret == 0) {
+        ret = wc_KyberKey_Init(type, kem, ssl->heap, INVALID_DEVID);
```
Comment (JacobBarthelmeh) [inline src/tls.c:7208]: Lets use the devID here, either from a dereference of 'ssl' or a function call to ` wolfSSL_CTX_GetDevId(WOLFSSL_CTX* ctx, WOLFSSL* ssl);`

File: src/ssl.c (hunk 1 above)
Comment (SparkiDev) [reply src/ssl.c:20702]: Done.

**Comment by SparkiDev** [comment]
@anhu 
Default --enable-kyber is for you to implement.
It will use the liboqs or the PMQ4(???) code to implement.
I suggest that enabling it will fail to configure if the dependent library is not available.

File: wolfssl/wolfcrypt/kyber.h (hunk 2 above)
Comment (SparkiDev) [reply wolfssl/wolfcrypt/kyber.h:1]: I made the 90s a configuration option.
You get one or the other but not both.

File: wolfssl/wolfcrypt/kyber.h (hunk 3 above)
Comment (SparkiDev) [reply wolfssl/wolfcrypt/kyber.h:161]: At config time you choose one set or the other.

File: configure.ac (hunk 4 above)
Comment (SparkiDev) [reply configure.ac:987]: We can't default to the wolfssl implementation.
Anthony will implement using liboqs or other APIs.

File: src/tls.c (hunk 5 above)
Comment (SparkiDev) [reply src/tls.c:7208]: Fixed to match what we do in TLSX_KeyShare_ProcessDh().

**Comment by SparkiDev** [comment]
KYBER_LEVEL1, KYBER_LEVEL3, KYBER_LEVEL5 added to kyber.h for user convenience.

File: configure.ac (hunk 4 above)
Comment (anhu) [reply configure.ac:987]: Ah, Ok.  I think I understand.  I'm currently working on this.

**Comment by anhu** [comment]
> Default --enable-kyber is for you to implement.
> It will use the liboqs or the PMQ4(???) code to implement.

I understand now and am working on this. 

> KYBER_LEVEL1, KYBER_LEVEL3, KYBER_LEVEL5 added to kyber.h for user convenience.

Excellent. Thank you much @SparkiDev !

File: src/tls.c [hunk 6]
```diff
@@ -8568,10 +8881,171 @@ static int TLSX_KeyShare_New(KeyShareEntry** list, int group, void *heap,
...
+
+    if (ret == 0) {
+        ret = wc_KyberKey_DecodePublicKey(kem, data + ecc_kse->pubKeyLen,
+            pubSz);
+    }
+    if (ret == 0) {
+        ret = wc_KyberKey_Encapsulate(kem, ciphertext + ecc_kse->pubKeyLen,
+            sharedSecret + outlen, ssl->rng);
+        if (ret != 0) {
+            WOLFSSL_MSG("OQS Encapsulation failure.");
```
Comment (anhu) [inline src/tls.c:9016]: This error message is misleading as OQS is not being used here. 

Copy paste error?  Might be elsewhere as well, but a quick search yields  nothing.

**Review by anhu — CHANGES_REQUESTED** [review]

File: src/tls.c (hunk 6 above)
Comment (SparkiDev) [reply src/tls.c:9018]: Copy-Paste error. Please consider changing in your next update.

**Comment by SparkiDev** [comment]
Fixed visibility issues when compiling shared library.

**Comment by anhu** [comment]
Decided to force push.  I'm already a contributor anyways.  @SparkiDev , sorry if I stepped on your toes here.

File: wolfcrypt/test/test.c [hunk 7]
```diff
@@ -29838,6 +29857,3977 @@ WOLFSSL_TEST_SUBROUTINE int ed448_test(void)
...
+#endif
+    if (ret != 0)
+        return -20000;
+
+    for (i = 0; i < (int)(sizeof(testData) / sizeof(*testData)); i++) {
+        ret = wc_KyberKey_Init(testData[i][0], &key, HEAP_HINT, INVALID_DEVID);
+        if (ret != 0)
+            return -20001 - i * 20;
+
+        ret = wc_KyberKey_MakeKey(&key, &rng);
```
Comment (JacobBarthelmeh) [inline wolfcrypt/test/test.c:33751]: Tests are reporting some leaking memory, please give it a once over with -fsanitize=address on a linux like system

./configure --enable-kyber --with-liboqs=/usr/local CC="clang -fsanitize=address" --enable-debug
make
./wolfcrypt/test/testwolfcrypt

```

Direct leak of 80 byte(s) in 1 object(s) allocated from:
    #0 0x4ded50 in malloc (/home/jak/Documents/wolfssl/wolfcrypt/test/.libs/testwolfcrypt+0x4ded50)
    #1 0x7f73a8561dbd in OQS_KEM_kyber_512_new (/home/jak/Documents/wolfssl/src/.libs/libwolfssl.so.35+0x3f3dbd)
    #2 0x534296 in kyber_test /home/jak/Documents/wolfssl/wolfcrypt/test/test.c:33751:15
    #3 0x519073 in wolfcrypt_test /home/jak/Documents/wolfssl/wolfcrypt/test/test.c:1360:17
    #4 0x535b47 in main /home/jak/Documents/wolfssl/wolfcrypt/test/test.c:1619:9
    #5 0x7f73a71bdc86 in __libc_start_main /build/glibc-CVJwZb/glibc-2.27/csu/../csu/libc-start.c:310
```

**Review by JacobBarthelmeh — CHANGES_REQUESTED** [review]
Looks great overall, ready to approve once the memory leak in the test case is investigated.

File: wolfssl/wolfcrypt/ext_kyber.h [hunk 8]
```diff
@@ -0,0 +1,62 @@
+/* ext_kyber.c
```
[the remark this answers is not in the document]
Comment (SparkiDev) [reply wolfssl/wolfcrypt/ext_kyber.h:1]: fixed

File: wolfcrypt/test/test.c (hunk 7 above)
Comment (SparkiDev) [reply wolfcrypt/test/test.c:33751]: fixed

**Review by JacobBarthelmeh — APPROVED** [review]

**Review by anhu — APPROVED** [review]

**Pull request merged**